repair/repair.cc

```diff
@@ -1192,19 +1199,31 @@ int repair_service::do_repair_start(sstring keyspace, std::unordered_map<sstring
         if (repair_tracker().is_aborted(id.uuid)) {
             throw std::runtime_error("aborted by user request");
         }
-
         for (auto shard : boost::irange(unsigned(0), smp::count)) {
             auto f = container().invoke_on(shard, [keyspace, table_ids, id, ranges, hints_batchlog_flushed,
-                    data_centers = options.data_centers, hosts = options.hosts, ignore_nodes] (repair_service& local_repair) mutable {
+                    data_centers = options.data_centers, hosts = options.hosts, ignore_nodes, tables_serialized = std::move(tables_serialized), shard] (repair_service& local_repair) mutable {
+                task_id local_id = task_id::create_random_id();
+                local_repair._task_manager_module->make_task<repair_task_impl>(shard, local_id, keyspace, tables_serialized, "", "", task_id{id.uuid}).get();
+                auto task = local_repair._task_manager_module->get_tasks()[local_id];
+                task->start();
+
                 local_repair.get_metrics().repair_total_ranges_sum += ranges.size();
-                auto ri = make_lw_shared<repair_info>(local_repair,
-                        std::move(keyspace), std::move(ranges), std::move(table_ids),
-                        id, std::move(data_centers), std::move(hosts), std::move(ignore_nodes), streaming::stream_reason::repair, id.uuid, hints_batchlog_flushed);
-                return repair_ranges(ri);
+                try {
+                    auto ri = make_lw_shared<repair_info>(local_repair,
+                            std::move(keyspace), std::move(ranges), std::move(table_ids),
+                            id, std::move(data_centers), std::move(hosts), std::move(ignore_nodes), streaming::stream_reason::repair, id.uuid, hints_batchlog_flushed);
+                    repair_ranges(ri).get();
+                    task->finish();
+                } catch(...) {
+                    auto ex = std::current_exception();
+                    task->finish_failed(ex);
+                    return make_exception_future(ex);
```

Comment:
when converting to a coroutine, you wouldn't be able to do that in the catch clause.
You can define `std::exception_ptr ex` outside try/catch, assign it here,
and after the catch block, do
```c++
if (ex) {
    co_await coroutine::return_exception_ptr(std::move(ex));
}
```